source/iotagent-architecture.rst

```diff
@@ -0,0 +1,543 @@
+   authorized device (which includes verifying whether a particular device is
+   authorized or not) should rely on public keys and/or signed certificates.
+#. Communication channel security: all messages exchanged with a physical
+   device should be encrypted using well-known cryptographic standards, such as
+   TLS. Any in-house security protocols should be avoided.
+#. Certificate revocation: the IoT agent should be able to discard any messages
+   from previously authorized device if its security data has been somehow
+   compromised. For instance, if the private key associated to a particular
+   device is leaked, then all its messages should be ignored as there is no
+   guarantee that they came from that device.
+
+Each of these aspects will be detailed in the following sections.
+
+Device identity
+---------------
+
+The device identity verification is the starting point when dealing with
+communication security. This validation will indicate to the IoT agent if the
+device that opened the connection is whoever it says it is. Furthermore, the
+IoT agent must, once this validation succeeds, check whether this device can
+connect to it by checking its ID. This section will show how to do that.
+
+For connection-oriented protocols, the IoT agent should only accept connections
+for devices that have a certificate that was signed by an authority that is
+trusted by dojot. Once this certificate is valid, device identity can be
+checked in two forms:
+
+- Device ID encoded in certificate: although this is a less-reliable mechanism,
+  it allows greater flexibility using many devices in a controlled deployment.
+  This is based on setting the common name (CN certificate field) as dojot
+  device ID. Therefore, IoT agent should check whether this device exists or
+  not and allow or deny the connection right away depending on this
+  verification. The weak points of this mechanisms is that the device
+  certificate must be signed by dojot's internal CA (once there is a procedure
+  to sign only one certificate per device) and, if this certificate is valid,
+  then its ID must also be valid. If any other CA is used, then this mechanism
+  has no valid use.
+
+- IoT agent has all valid certificates: if an administrator wants to use an
+  external CA to sign all device certificates, then there is no actual control
+  of which device ID was used to generate a particular certificate. Therefore,
+  IoT agent must have all valid certificates properly mapped onto a device list
+  - this will guarantee that only one certificate is allowed to a particular
+  device and vice-versa.
+
+Using the first mechanism, the device (or an operator configuring a device for
+the first time) must call dojot CA to generate a signed certificate for itself.
+There is no further action for IoT agent to take as long as dojot CA is used.
+
+The second mechanism, however, requires that an IoT agent offer methods to
```

Comment:
I don't get this paragraph, what does it mean for its certificates to be managed?

Reply:
The IoT agent must provide ways to, at least, add and revoke certificates.